dmoj/cptbox/ptdebug_x64.cpp

```diff
@@ -2,96 +2,136 @@
 #define _BSD_SOURCE
 #include "ptbox.h"
 
-#ifdef HAS_DEBUGGER_X64
+#ifdef __amd64__
+#include <asm/unistd.h>
+#include <elf.h>
 #include <errno.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <sys/ptrace.h>
+#include <sys/uio.h>
 
-#define R15 0
-#define R14 1
-#define R13 2
-#define R12 3
-#define RBP 4
-#define RBX 5
-#define R11 6
-#define R10 7
-#define R9 8
-#define R8 9
-#define RAX 10
-#define RCX 11
-#define RDX 12
-#define RSI 13
-#define RDI 14
-#define ORIG_RAX 15
-#define RIP 16
-#define CS 17
-#define EFLAGS 18
-#define RSP 19
-#define SS 20
-#define FS_BASE 21
-#define GS_BASE 22
-#define DS 23
-#define ES 24
-#define FS 25
-#define GS 26
+bool pt_debugger::supports_abi(int abi) {
+    switch (abi) {
+        case PTBOX_ABI_X86:
+        case PTBOX_ABI_X64:
+        case PTBOX_ABI_X32:
+            return true;
+    }
+    return false;
+}
 
-int pt_debugger_x64::syscall() {
-    return (int) peek_reg(ORIG_RAX);
+void pt_debugger::pre_syscall() {
+    struct iovec iovec;
+    iovec.iov_base = &regs;
+    iovec.iov_len = sizeof regs;
+    if (ptrace(PTRACE_GETREGSET, tid, NT_PRSTATUS, &iovec)) {
+        perror("ptrace(PTRACE_GETREGSET)");
+        abi_ = PTBOX_ABI_INVALID;
+    } else {
+        if (iovec.iov_len == sizeof regs.x86) {
+            abi_ = PTBOX_ABI_X86;
+        } else if (regs.x64.orig_rax & __X32_SYSCALL_BIT) {
+            abi_ = PTBOX_ABI_X32;
+        } else {
+            abi_ = PTBOX_ABI_X64;
+        }
+        regs_changed = false;
+    }
 }
 
-void pt_debugger_x64::syscall(int id) {
-#if PTBOX_FREEBSD
-    poke_reg(RAX, id);
-#else
-    poke_reg(ORIG_RAX, id);
-#endif
+void pt_debugger::post_syscall() {
+    if (!regs_changed || abi_ == PTBOX_ABI_INVALID)
+        return;
+
+    struct iovec iovec;
+    iovec.iov_base = &regs;
+    iovec.iov_len = abi_ == PTBOX_ABI_X86 ? sizeof regs.x86 : sizeof regs.x64;
+    if (ptrace(PTRACE_SETREGSET, tid, NT_PRSTATUS, &iovec)) {
+        perror("ptrace(PTRACE_SETREGSET)");
```

Comment:
We could've failed to change a syscall to a no-op here, is this ok?

Reply:
Not quite, but it's no worse than before, when updating every register could fail. We need to think of a way to report the errors, but that's probably for later.